<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>(m)RNA sequencing</title>
    <meta charset="utf-8" />
    <meta name="author" content="Konstantin Zaitsev" />
    <link rel="stylesheet" href="..\libs\itmo.css" type="text/css" />
    <link rel="stylesheet" href="..\libs\itmo-fonts.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# (m)RNA sequencing
### Konstantin Zaitsev
### October 31<sup>st</sup>, 2020

---


class: center, middle

# (m)RNA sequencing

---

## Why we measure RNA abundance

&lt;img src="dogma.png" width="55%"&gt;

&lt;div class="my-footer"&gt;&lt;span&gt;Taken from https://www.ncbi.nlm.nih.gov/pubmed/26259698&lt;/span&gt;&lt;/div&gt; 

---

## Gene structure

.center[
&lt;img src="gene_structure.jpg" width="70%"&gt;
]


&lt;div class="my-footer"&gt;&lt;span&gt;Taken from http://nitro.biosci.arizona.edu/courses/EEB600A-2003/lectures/lecture24/lecture24.html&lt;/span&gt;&lt;/div&gt; 

---

## Gene structure: translation

.center[
&lt;img src="translation.png" width="90%"&gt;
]


&lt;div class="my-footer"&gt;&lt;span&gt;Taken from https://www.ncbi.nlm.nih.gov/pubmed/23810193&lt;/span&gt;&lt;/div&gt; 

---


## Gene structure: translation

.center[
&lt;img src="translation.png" width="90%"&gt;
]

.center[
&lt;img src="mrna.png" width="90%"&gt;
]


&lt;div class="my-footer"&gt;&lt;span&gt;Taken from https://www.ncbi.nlm.nih.gov/pubmed/23810193&lt;/span&gt;&lt;/div&gt; 

---
## Some genes are simple

.center[
&lt;img src="cd14.png" width="90%"&gt;
]

---

## Many genes have multiple isoforms

.center[
&lt;img src="irf1.png" width="90%"&gt;
]

---


## Most of the RNA is ribosomal RNA

.center[
&lt;img src="rrna.png" width="65%"&gt;
]


&lt;div class="my-footer"&gt;&lt;span&gt;Taken from http://hpst.cz/sites/default/files/attachments/5991-5938en.pdf&lt;/span&gt;&lt;/div&gt; 
---


## Strategies to capture RNA

|Strategy         |	Type of RNA	    | Ribosomal RNA content	 | Unprocessed RNA content	| Genomic DNA content | Isolation method                                             |
|:----------------|:---------------:|:----------------------:|:------------------------:|:-------------------:|:------------------------------------------------------------:|
|Total  RNA       |All	            | High                   |High	                    |High	                |   None                                                       |
|PolyA selection	|Coding           |	Low	                   |Low	                      |Low	                |Hybridization with poly(dT) oligomers                         |
|rRNA depletion	  |Coding, noncoding|	Low	                   |High	                    |High	                |Removal of oligomers complementary to rRNA                    |
|RNA capture	    |Targeted	        | Low	                   |Moderate	                |Low	                |Hybridization with probes complementary to desired transcripts|

---

## Comparison

.center[
&lt;img src="isolation_comparison.png" width="60%"&gt;
]


&lt;div class="my-footer"&gt;&lt;span&gt;Taken from https://bmcgenomics.biomedcentral.com/track/pdf/10.1186/1471-2164-15-419&lt;/span&gt;&lt;/div&gt; 
---

## What is RNA-seq

.center[
&lt;img src="rna_seq.png" width="100%"&gt;
]


&lt;div class="my-footer"&gt;&lt;span&gt;Taken from https://www.biostat.wisc.edu/bmi776/lectures/rnaseq.pdf&lt;/span&gt;&lt;/div&gt; 
---

## RNA-seq

RNA-seq can be done very differently:

* 3' RNA-seq is good for measuring expression
* Full-length (paired-end) RNA-seq is good for detecting isoforms
* You can target specific sequences

Depending on what you want to do protocol steps and their order (such as fragmentation) might vary


---

## RNA-seq for gene expression

.center[
&lt;img src="3prime.png" width="65%"&gt;
]


---

## Traditional 3' RNA-seq pipeline

.center[
&lt;img src="3prime_pipeline.png" width="65%"&gt;
]


---

## Traditional 3' RNA-seq pipeline

Once RNA is fragmented

.center[
&lt;img src="library_prep.png" width="70%"&gt;
]


---

## 3' vs full-length RNA-seq


.center[
&lt;img src="3_vs_full.png" width="50%"&gt;
]

---

class: center, middle

# Any questions so far?

---

## Processing RNA-seq: two cases

* Model organism with a good reference genome (human, mouse, arabidopsis, yeast)
* Non-model organism :(

.center[
&lt;img src="alien.jpg" width="20%"&gt;
]



---

## Basic RNA-seq analysis pipeline

* Usually we can live without clipping/trimming
* Alignment: raw reads (fastq) + reference genome -&gt; bam
* Quantification: bam (aligned reads) + reference genome -&gt; expression table
* Secondary analysis: well, that part you know already

---

## Refernce genomes

* Human: chr1-chr22, chrX, chrY, chrM: 
  * 3235 Mb, 20412 genes
* Mouse: Chr1-chr19, chrX, chrY, chrM: 
  * 2318 Mb, 23,139 genes
* Assembly is mostly complete: but not 100%, there are unplaced scaffolds and gaps
* There are pseudogenes, lncRNA, rRNA and other stuff in a genome, which are usually present in the genome annotation

---

## Popular genome assemblies

* Human:
  * UCSC notation (hg19, hg38)
  * Genome reference consortium notation (major: GRCh37, minor: GRCh38.p7)
  * 1000 genomes notation (b37)
* Mouse - same (mm10, GRCm37)

---

## Gencode: my personal preference

* https://www.gencodegenes.org/
* I usually use primary assembly
.center[
&lt;img src="primary.png" width="100%"&gt;
]

---

## Gencode: my personal preference

.center[
&lt;img src="release_history.png" width="100%"&gt;
]

---

## What is reference

Usually reference is just two files:

* Sequence fasta file
* GTF/GFF annotation

---

## Sequence Fasta files

```
[kzaytsev@login02 hg38]$ head -4 GRCh38.primary_assembly.genome.fa
&gt;chr1 1
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
[kzaytsev@login02 hg38]$ head -500 GRCh38.primary_assembly.genome.fa | tail -4
GGCTCCTCCGGGCAGCCCTCGCCCGGGGTGCGCCCCGGGGCAGGACCCCCAGCCCACGCC
CAGGGCCCGCCCCTGCCCTCCAGCCCTACGCCTTGACCCGCTTTCCTGCGTCTCTCAGCC
TACCTGACCTTGTCTTTACCTCTGTGGGCAGCTCCCTTGTGATCTGCTTAGTTCCCACCC
CCCTTTAAGAATTCAATAGAGAAGCCAGACGCAAAACTACAGATATCGTATGAGTCCAGT
```

---

## GTF Annotation file


```
[kzaytsev@login02 hg38]$ head gencode.v28.primary_assembly.annotation.gtf
##description: evidence-based annotation of the human genome (GRCh38), version 28 (Ensembl 92)
##provider: GENCODE
##contact: gencode-help@ebi.ac.uk
##format: gtf
##date: 2018-03-23
chr1    HAVANA  gene    11869   14409   .       +       .       gene_id "ENSG00000223972.5"; gene_type "transcribed_unprocessed_pseudogene"; gene_name "DDX11L1"; level 2; havana_gene "OTTHUMG00000000961.2";
chr1    HAVANA  transcript      11869   14409   .       +       .       gene_id "ENSG00000223972.5"; transcript_id "ENST00000456328.2"; gene_type "transcribed_unprocessed_pseudogene"; gene_name "DDX11L1"; transcript_type "processed_transcript"; transcript_name "RP11-34P13.1-002"; level 2; transcript_support_level "1"; tag "basic"; havana_gene "OTTHUMG00000000961.2"; havana_transcript "OTTHUMT00000362751.1";
chr1    HAVANA  exon    11869   12227   .       +       .       gene_id "ENSG00000223972.5"; transcript_id "ENST00000456328.2"; gene_type "transcribed_unprocessed_pseudogene"; gene_name "DDX11L1"; transcript_type "processed_transcript"; transcript_name "RP11-34P13.1-002"; exon_number 1; exon_id "ENSE00002234944.1"; level 2; transcript_support_level "1"; tag "basic"; havana_gene "OTTHUMG00000000961.2"; havana_transcript "OTTHUMT00000362751.1";
chr1    HAVANA  exon    12613   12721   .       +       .       gene_id "ENSG00000223972.5"; transcript_id "ENST00000456328.2"; gene_type "transcribed_unprocessed_pseudogene"; gene_name "DDX11L1"; transcript_type "processed_transcript"; transcript_name "RP11-34P13.1-002"; exon_number 2; exon_id "ENSE00003582793.1"; level 2; transcript_support_level "1"; tag "basic"; havana_gene "OTTHUMG00000000961.2"; havana_transcript "OTTHUMT00000362751.1";
chr1    HAVANA  exon    13221   14409   .       +       .       gene_id "ENSG00000223972.5"; transcript_id "ENST00000456328.2"; gene_type "transcribed_unprocessed_pseudogene"; gene_name "DDX11L1"; transcript_type "processed_transcript"; transcript_name "RP11-34P13.1-002"; exon_number 3; exon_id "ENSE00002312635.1"; level 2; transcript_support_level "1"; tag "basic"; havana_gene "OTTHUMG00000000961.2"; havana_transcript "OTTHUMT00000362751.1";
```
---

## GTF Annotation file

.center[
&lt;img src="gtf.png" width="100%"&gt;
]


---


## Fastq files

Raw sequencing files are FastQ files files

```
[kzaytsev@login02 askembefore]$ zcat foam_01_TACAGAC.fq.gz | head -12
@J00136:207:HLNWCBBXX:2:1101:1194:1068 1:N:0:NACAGAC
NGGGCGGATACAGAGAATAGCTTTTAAAGTGTTCTATGTTCCCAGGGGCA
+
#AAFFJJJJJJJJFJJFJJJJJJFFJJJF&lt;JJ-FFF&lt;-&lt;FFJJFJJJJJF
@J00136:207:HLNWCBBXX:2:1101:1215:1068 1:N:0:NACAGAC
NTGCTCATCCTGGCTTACGTCATTGGGTGTGTGTTCGTGTGGGAAAAGGG
+
#&lt;-A&lt;AFAJ&lt;&lt;&lt;FJJFJFF-FA7AFJF-7-7-&lt;--&lt;&lt;-&lt;JJ-7A-&lt;7A-A
@J00136:207:HLNWCBBXX:2:1101:1357:1068 1:N:0:NACAGAC
NGTGGCTCTGGAGGGCGTAGGCCACTTCTTCCGCGAAGTGGCCGAGGAGA
+
#AAFFFFJJFJFJJJAFJJ&lt;FFJFJJJ&lt;&lt;FJJ&lt;JJJF-&lt;AAFJFJ-JJJJ
```

---

## Fastq files

Raw sequencing files are FastQ files files:

* Header: instrument name, run number, flowcell ID, lane number, tile number, X coord, Y coord. Read number, Y-filtered/N-not, control number, sometimes demultiplex barcode.
* Sequence
* +
* Quality string


```
[kzaytsev@login02 askembefore]$ zcat foam_01_TACAGAC.fq.gz | head -4
@J00136:207:HLNWCBBXX:2:1101:1194:1068 1:N:0:NACAGAC
NGGGCGGATACAGAGAATAGCTTTTAAAGTGTTCTATGTTCCCAGGGGCA
+
#AAFFJJJJJJJJFJJFJJJJJJFFJJJF&lt;JJ-FFF&lt;-&lt;FFJJFJJJJJF
```
---


## Alignment

.small[
.pull-left[Alignment:
* Bowtie/Bowtie2
* Hisat/Hisat2
* STAR

Counting:
* featureCounts
* htseq
* mmquant
* RSEM
]
]

.pull-right[
&lt;img src="gene_mapping.png" width="80%"&gt;
]

---

## Alignment

Most of the aligners align to genome, not transcriptome:

* Allows us to identify intergenic reads, DNA reads and other weird stuff

Alignment will produce SAM/BAM file:

* Contains mapping of every read (or read pair)

---

## Alignment

Read name, bitwise flag, sequence name (CHROMOSOME), mapping position, mapping quality, CIGAR String and other stuff

Might differ from aligner to aligner, read specification

```
[kzaytsev@login02 askembefore]$ samtools view total_foam.bam | head -4
K00193:243:HN5TKBBXX:6:1109:8937:17368  16      1       3013057 255     98M     *       0       0       GTGATTTATTCTTCTGTATCTGGTTAGTGTTTATTGATTTAACTCATCCATACTACTTAGGTCTTCCATATGTCTTTGGTCTATATGTCTGTTTTTGT JJJJAAFJFJFFFFJJFJJJJJJFJJJJJJJJFJJJ&lt;FFJJJJJJJJFJJJJFJJJJJJJJFJJJJJJJJJJFJJJJJJJJJJJJFJJFJJJJFFF&lt;A       NH:i:1  HI:i:1  AS:i:96 nM:i:0  RE:A:I  CR:Z:TACCTATTCAGCAACT   CY:Z:AAFFFJJJJJJJJJJJ   CB:Z:TACCTATTCAGCAACT-1     UR:Z:AGTCAATCAT UY:Z:JJJJJJJJJJ UB:Z:AGTCAATCAT BC:Z:CTGCGGCT   QT:Z:AAAFFFJJ   RG:Z:M_FN-Ao_FC-Ao_FC:MissingLibrary:1:HN5TKBBXX:6
K00193:243:HN5TKBBXX:6:2101:9831:28094  16      1       3013057 255     98M     *       0       0       GTGANTTATTCTTCTGTATCTGGTTAGTGTTTATTGATTTAACTCATCCATACTACTTAGGTCTTCCATATGTCTTTGGTCTATATGTCTGTTTTTGT JJJJ#FJJJJJJJJFFF7JJJJJJJJJJJJJJFJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJJJFFF&lt;A       NH:i:1  HI:i:1  AS:i:95 nM:i:0  RE:A:I  CR:Z:TACCTATTCAGCAACT   CY:Z:AAFFFJJJJJJJJJJJ   CB:Z:TACCTATTCAGCAACT-1     UR:Z:AGTCAATCAT UY:Z:JJJJJJJJJJ UB:Z:AGTCAATCAT BC:Z:AGAAACTC   QT:Z:AAFFFJJJ   RG:Z:M_FN-Ao_FC-Ao_FC:MissingLibrary:1:HN5TKBBXX:6
K00193:243:HN5TKBBXX:6:2101:9475:28885  16      1       3013057 255     98M     *       0       0       GTGANTTATTCTTCTGTATCTGGTTAGTGTTTATTGATTTAACTCATCCATACTACTTAGGTCTTCCATATGTCTTTGGTCTATATGTCTGTTTTTGT JJJA#JJJJJJJJFFJF7JJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFFF&lt;A       NH:i:1  HI:i:1  AS:i:95 nM:i:0  RE:A:I  CR:Z:TACCTATTCAGCAACT   CY:Z:AAFFFJJJJJJJJJJJ   CB:Z:TACCTATTCAGCAACT-1     UR:Z:AGTCAATCAT UY:Z:JJJJJJJJJJ UB:Z:AGTCAATCAT BC:Z:AGAAACTC   QT:Z:AAFFFJJJ   RG:Z:M_FN-Ao_FC-Ao_FC:MissingLibrary:1:HN5TKBBXX:6
K00193:243:HN5TKBBXX:6:2204:2443:3723   272     1       3014816 0       7S91M   *       0       0       CCCCATCCCCCCCCCCCCAATCCCCTCCCCCCCCGCTCGCCCTCTTTGGCCCTGGCGTTCCCCTGTACTGGGGCTTATAACGTTTGCTCGTCCTATGG 77-&lt;A7-7-7-A777F77--&lt;7--A7-A7----7--77-AJJJAJJJJFFFJJJF7FF77--FA&lt;&lt;7AJFAF&lt;7-7-&lt;---A&lt;AAF7--FF&lt;&lt;--F&lt;A       NH:i:5  HI:i:1  AS:i:69 nM:i:10 RE:A:I  CR:Z:GTCGGGTGTCTAAACC   CY:Z:AAAFAAA&lt;-AAFJJJJ   CB:Z:GTCGGGTGTCTAAACC-1     UR:Z:GTGACCGTCT UY:Z:77AAJJA-77 UB:Z:GTGACCGTCT BC:Z:AGAAACTC   QT:Z:AAAFFJ&lt;F   RG:Z:M_FN-Ao_FC-Ao_FC:MissingLibrary:1:HN5TKBBXX:6
```

.center[
&lt;img src="alignment_screen.png" width="100%"&gt;
]

---

## Samtools: working with alignment files

If you need to work with alignment files, you most likely will work with samtools:

* `samtools view`
* `samtools sort`
* `samtools index`

---

## Read coverage

Once we have bam file (usually sorted and indexed) we can create coverage tracks to be visualized in genome browser. We prefer bigwig format

* Usual thing to use in this case is https://deeptools.readthedocs.io/en/develop/content/tools/bamCoverage.html

---

## Quality controls

* FastQ files: FastQC
* Alignment: (RSeQC) http://rseqc.sourceforge.net/
  * infer_experiment.py
  * read_distribution.py
  * geneBody_coverage.py
* Picard CollectRnaSeqMetrics
* MultiQC to generate summary report for all these tools

---

## Libraries:

* Paired-end, single-end
* Stranded, Unstranded
* 3', 5', full-length

---

## Quantification (counting): featureCounts

.pull-left[
* For stranded experiment we can identify which of two genes is expressed if they are on opposite strands
* For non-stranded - no
* Different mappers act differently with multimapped or ambiguous reads
]

.pull-right[
&lt;img src="mapping_happens.png" width="100%"&gt;
]

---

## Kallisto: pseudoaligner

.pull-left[
* Pseudoalligner: we build T-DBG (transcriptome de Bruijn graph) with relatively small k
* We calculate *k-equivalence* classes
* We hash mapping of k-mers to classes 
* No bam
* Super fast
]

.pull-right[
&lt;img src="kallisto.png" width="80%"&gt;
]

&lt;div class="my-footer"&gt;&lt;span&gt;Taken from https://www.ncbi.nlm.nih.gov/pubmed/27043002&lt;/span&gt;&lt;/div&gt; 

---

## Kallisto: pseudoaligner

.pull-left[
* Pseudoalligner: we build T-DBG (transcriptome de Bruijn graph) with relatively small k
* We calculate *k-equivalence* classes
* We hash mapping of k-mers to classes 
* No bam
* **Super fast**
]

.pull-right[
&lt;img src="super_fast.png" width="100%"&gt;
]

&lt;div class="my-footer"&gt;&lt;span&gt;Taken from https://www.ncbi.nlm.nih.gov/pubmed/27043002&lt;/span&gt;&lt;/div&gt; 

---

## Once we have counts with table

Analysis in R is all the same but done with DESeq2 package:
* Different normalizations for RNA-seq
* Different statistical model for counts (negative binomial distributions)

---

## Deseq2 installing


```r
if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
if (!requireNamespace("ggplot2", quietly = TRUE)) install.packages("ggplot2")
if (!requireNamespace("ggrepel", quietly = TRUE)) install.packages("ggrepel")
if (!requireNamespace("dplyr", quietly = TRUE)) install.packages("dplyr")
if (!requireNamespace("DESeq2", quietly = TRUE)) BiocManager::install("DESeq2")
if (!requireNamespace("apeglm", quietly = TRUE)) BiocManager::install("apeglm")
if (!requireNamespace("org.Mm.eg.db", quietly = TRUE)) BiocManager::install("org.Mm.eg.db")
```

---
## Deseq2 examples:


```r
library(ggplot2)
library(DESeq2)
library(apeglm)
library(ggrepel)
library(dplyr)
library(org.Mm.eg.db)


countFiles &lt;- list.files("GSE116239_RAW", full.names = T)
countFiles
```

```
## [1] "GSE116239_RAW/GSM3215429_foam_01_gene_counts.txt"    
## [2] "GSE116239_RAW/GSM3215430_foam_02_gene_counts.txt"    
## [3] "GSE116239_RAW/GSM3215431_foam_03_gene_counts.txt"    
## [4] "GSE116239_RAW/GSM3215432_non_foam_01_gene_counts.txt"
## [5] "GSE116239_RAW/GSM3215433_non_foam_02_gene_counts.txt"
## [6] "GSE116239_RAW/GSM3215434_non_foam_03_gene_counts.txt"
```

---
## Deseq2 examples:

.tiny[

```r
counts &lt;- lapply(countFiles, function(countsFile) {
  read.table(countsFile, sep="\t", header=1, row.names = 1, stringsAsFactors = F, comment.char = "")
})

head(counts[[1]])
```

```
##                              Chr
## ENSMUSG00000090025             1
## ENSMUSG00000064842             1
## ENSMUSG00000051951 1;1;1;1;1;1;1
## ENSMUSG00000089699           1;1
## ENSMUSG00000088333             1
## ENSMUSG00000025900       1;1;1;1
##                                                                      Start
## ENSMUSG00000090025                                                 3054233
## ENSMUSG00000064842                                                 3102016
## ENSMUSG00000051951 3205901;3206523;3213439;3213609;3214482;3421702;3670552
## ENSMUSG00000089699                                         3466587;3513405
## ENSMUSG00000088333                                                 3783876
## ENSMUSG00000025900                         4343507;4351910;4352202;4360200
##                                                                        End
## ENSMUSG00000090025                                                 3054733
## ENSMUSG00000064842                                                 3102125
## ENSMUSG00000051951 3207317;3207317;3215632;3216344;3216968;3421901;3671498
## ENSMUSG00000089699                                         3466687;3513553
## ENSMUSG00000088333                                                 3783933
## ENSMUSG00000025900                         4350091;4352081;4352837;4360314
##                           Strand Length Count CPM RPKM
## ENSMUSG00000090025             +    501     0   0    0
## ENSMUSG00000064842             +    110     0   0    0
## ENSMUSG00000051951 -;-;-;-;-;-;-   6094     0   0    0
## ENSMUSG00000089699           +;+    250     0   0    0
## ENSMUSG00000088333             -     58     0   0    0
## ENSMUSG00000025900       -;-;-;-   7508     0   0    0
```
]


---

## Deseq2 examples:


```r
counts &lt;- lapply(counts, function(countsTable) countsTable[, "Count", drop=F])
counts &lt;- do.call(cbind, counts)
colnames(counts) &lt;- gsub(".*(GSM\\d+).*", "\\1", countFiles)
head(counts)
```

```
##                    GSM3215429 GSM3215430 GSM3215431 GSM3215432 GSM3215433
## ENSMUSG00000090025          0          0          0          2          0
## ENSMUSG00000064842          0          0          0          0          0
## ENSMUSG00000051951          0          0          0          4          0
## ENSMUSG00000089699          0          0          0          0          0
## ENSMUSG00000088333          0          0          0          0          0
## ENSMUSG00000025900          0          2          0          2          1
##                    GSM3215434
## ENSMUSG00000090025          0
## ENSMUSG00000064842          0
## ENSMUSG00000051951          1
## ENSMUSG00000089699          0
## ENSMUSG00000088333          0
## ENSMUSG00000025900          0
```

---
## Deseq2 examples:


```r
coldata &lt;- data.frame(
  gsm=gsub(".*(GSM\\d+).*", "\\1", countFiles),
  foam=gsub(".*(GSM\\d+)_(foam|non_foam).*", "\\2", countFiles),
  row.names =gsub(".*(GSM\\d+).*", "\\1", countFiles)
)

coldata
```

```
##                   gsm     foam
## GSM3215429 GSM3215429     foam
## GSM3215430 GSM3215430     foam
## GSM3215431 GSM3215431     foam
## GSM3215432 GSM3215432 non_foam
## GSM3215433 GSM3215433 non_foam
## GSM3215434 GSM3215434 non_foam
```

---
## Deseq2 examples:


```r
# only keeping genes that have at least 10 reads
dds &lt;- DESeqDataSetFromMatrix(countData = counts[rowSums(counts) &gt; 10, ],
                              colData = coldata,
                              design= ~ foam)
```

```
## Warning in DESeqDataSet(se, design = design, ignoreRank): some variables in
## design formula are characters, converting to factors
```

```r
dds &lt;- DESeq(dds)
resultsNames(dds) # lists the coefficients
```

```
## [1] "Intercept"             "foam_non_foam_vs_foam"
```

---

## Deseq2 examples: PCA


```r
vst &lt;- varianceStabilizingTransformation(dds)
plotPCA(vst, intgroup=c("foam"))
```

![](rna_seq_files/figure-html/unnamed-chunk-7-1.svg)&lt;!-- --&gt;

---

## Deseq2 examples: DE


```r
res &lt;- lfcShrink(dds, coef="foam_non_foam_vs_foam", type="apeglm")
head(res)
```

```
## log2 fold change (MAP): foam non foam vs foam 
## Wald test p-value: foam non foam vs foam 
## DataFrame with 6 rows and 5 columns
##                      baseMean log2FoldChange     lfcSE      pvalue      padj
##                     &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt;
## ENSMUSG00000033845 1629.91001      0.0502838  0.225462 0.801100279 0.9381684
## ENSMUSG00000025903 3061.43086      0.1598131  0.236935 0.436676572 0.7391732
## ENSMUSG00000033813  394.35179      0.3165529  0.269558 0.156188811 0.4372599
## ENSMUSG00000002459    1.83531      0.0200497  0.461238 0.770906520        NA
## ENSMUSG00000033793 4859.39402     -0.7469256  0.210676 0.000103276 0.0022901
## ENSMUSG00000025905    2.77100      0.0807188  0.468554 0.345656766        NA
```

---

## Mapping genes


```r
keytypes(org.Mm.eg.db)
```

```
##  [1] "ACCNUM"       "ALIAS"        "ENSEMBL"      "ENSEMBLPROT"  "ENSEMBLTRANS"
##  [6] "ENTREZID"     "ENZYME"       "EVIDENCE"     "EVIDENCEALL"  "GENENAME"    
## [11] "GO"           "GOALL"        "IPI"          "MGI"          "ONTOLOGY"    
## [16] "ONTOLOGYALL"  "PATH"         "PFAM"         "PMID"         "PROSITE"     
## [21] "REFSEQ"       "SYMBOL"       "UNIGENE"      "UNIPROT"
```

```r
res$Gene.symbol &lt;- mapIds(org.Mm.eg.db, rownames(res), column="SYMBOL", "ENSEMBL")
```

---

## Deseq2 examples: volcano


```r
res &lt;- as.data.frame(res)
ggplot(res, aes(x=log2FoldChange, y=-log10(padj), color=padj &lt; 0.05)) +
  geom_point() + theme_bw() + scale_color_manual(values=c("black", "red")) +
  geom_text_repel(data=res %&gt;% dplyr::filter(padj &lt; 1e-20), aes(label=Gene.symbol, color=NULL))
```

![](rna_seq_files/figure-html/unnamed-chunk-10-1.svg)&lt;!-- --&gt;

---
class: center, middle

# Questions?
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"ratio": "16:9",
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
