---
title: "Gene expression 2020: syllabus"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Before we start

In this course, we (Konstantin Zaitsev and Alexander Tkachenko) will try to explain most of the essential things you need to now about gene expression and how to make the best of gene expression studies. We will go from explaining ways to capture RNA and measure RNA abundance to discussing how to properly design gene expression study and what the best practices are.

Authors of this course:

1. Konstantin Zaitsev, https://www.researchgate.net/profile/Konstantin_Zaitsev, https://stepik.org/course/512/
2. Alexander Tkachenko, https://publons.com/researcher/3041889/alexander-tkachenko/, https://stepik.org/course/94/


Most of the material updates will be published here.

Disclaimers:

1. All the course materials will be in English
2. By "gene expression" we almost always mean "RNA expression" if not said otherwise

## Syllabus

+ Central dogma of molecular biology, structure of gene, types of RNA, structure of RNA, transcription, reverse transcription, qPCR, FISH, <a href="https://konsolerr.github.io/gene_expression_2019/facs" target="_blank">FACS</a>
+ Microarray: quantification, normalization, basic analysis
+ RNA-seq: alignment, quantification, QC, normalization, basic analysis
+ Overall quality control: PCA, clustering, outlier detection
+ Overall quality control: batch correction
+ Differential expression (DE): limma for microarray, Deseq2 for RNA-seq
+ Downstream analysis: pathway/gene set enrichment analysis
+ Downstream analysis: gene expression deconvolution
+ Transcriptome assembly, functional annotation
+ Single-cell transcriptomics: Seurat basic analysis
+ Single-cell transcriptomics: Trajectory analysis, RNA velocity, optimal transport
+ Visual data exploration: phantasus, JBR genome browser
+ Experimental design of gene expression study

## Prerequisites

Most of the lessons will be in format of seminars: lectures and practical part with your laptops. Prior the start of the course, you are required to install latest version of R (https://cran.r-project.org/bin/) and Rstudio (https://www.rstudio.com/) to your laptop. Most of the practices we will be working in R and only for several we will require a command line.

**N.B.** all the code on the slides is built with R v.4.0.2. Usage of some of the packages might differ from version to version, so you might have to run the code differently if you are using different version of R.

## Homeworks 

There will be 4 projects homeworks during this course:

1. Microarray dataset: processing, QC, analysis, DE
2. RNA-seq dataset: processing (from fastq), QC, analysis, DE, downstream analysis (pathways)
3. Transcriptome assembly and functional annotation
4. Single-cell RNA-seq dataset: analysis with Seurat


## Below we will put links to slides and other useful links

### Lecture materials

1. Lecture 1: <a href="https://konsolerr.github.io/gene_expression_2020/intro" target="_blank">introduction</a> and <a href="https://konsolerr.github.io/gene_expression_2020/facs" target="_blank">FACS</a>
2. Lecture 2: <a href="https://konsolerr.github.io/gene_expression_2020/transcription/Transcription.pdf" target="_blank">transcription</a>
3. Lecture 3: Microarrays <a href="https://konsolerr.github.io/gene_expression_2020/microarray/microarray.html" target="_blank">part 1</a> and <a href="https://konsolerr.github.io/gene_expression_2020/microarray/microarray2.html" target="_blank">part 2</a>
4. Lecture 4: secondary analysis in gene expression studies. <a href="https://konsolerr.github.io/gene_expression_2020/secondary_analysis/" target="_blank">Slides</a> and <a href="https://konsolerr.github.io/gene_expression_2020/secondary_analysis/secondary_analysis_materials.zip" target="_blank">materials</a>.
5. Lecture 5: RNA-seq <a href="https://konsolerr.github.io/gene_expression_2020/rna-seq" target="_blank">slides</a> and <a href="https://konsolerr.github.io/gene_expression_2020/rna-seq/rna_seq_materials.zip" target="_blank">materials</a>.
6. Lecture 6: RNA-seq normalization and DE <a href="https://konsolerr.github.io/gene_expression_2020/normalization/normalization" target="_blank">slides</a> and <a href="https://konsolerr.github.io/gene_expression_2020/normalization/normalization_materials.zip" target="_blank">materials</a>.
7. Lecture 7: Transcriptome assembly and annotation. <a href="https://konsolerr.github.io/gene_expression_2020/transcriptome_assembly.pdf" target="_blank">Slides</a>
8. Lecture 8: scRNA-seq <a href="https://konsolerr.github.io/gene_expression_2020/scRNA-seq" target="_blank">slides</a> and <a href="https://konsolerr.github.io/gene_expression_2020/scRNA-seq/materials.zip" target="_blank">materials</a>. Brief lecture about trajectory methods for scRNA-seq: <a href="https://konsolerr.github.io/gene_expression_2019/scRNA-seq/trajectory.pdf" target="_blank">slides</a>
9. Lecture 9: Visual data exploration. Phantasus, single-cell navigator, JBR genome browser: <a href="https://konsolerr.github.io/gene_expression_2020/visual_analysis_expression.pdf" target="_blank">slides</a>
10. Lecture 10: Experimental design of RNA-seq studies: <a href="https://konsolerr.github.io/gene_expression_2020/experimental_design_in_rna_seq.pdf" target="_blank">slides</a>.

Random materials:

1. Good chapter on SVD: <a href="https://www.cs.cmu.edu/~venkatg/teaching/CStheory-infoage/book-chapter-4.pdf" target="_blank">https://www.cs.cmu.edu/~venkatg/teaching/CStheory-infoage/book-chapter-4.pdf</a>
2. Good interactive material "Understanding UMAP": <a href="https://pair-code.github.io/understanding-umap/" target="_blank">https://pair-code.github.io/understanding-umap/</a>
3. How much scRNA-seq library prep costs: https://satijalab.org/costpercell
4. About distributions behind box plots: https://www.autodeskresearch.com/node/3699


## Homeworks

### Homework 1: deadline November 21st

The first homework is to analyze a microarray gene expression dataset. All the datasets will be available by GEO id, so you can find them at GEO omnibus and also use `GEOquery::getGEO("GSEXXXXX")` to get the expression data into R session.

Basically you needed to perform basic steps for a dataset: normalize and make sense of the PCA plot. Then you need to design differential expression to address the general question for your homework. Once you perform DE I want you to highlight genes that are different and perform pathway enrichment.

Datasets IDs are listed in the table below:

<a href="https://docs.google.com/spreadsheets/d/1DpZUgJ3iEkKcFCNYv5PHz1bEcltX4UbQtB-uFTCc6S8/edit?usp=sharing" target="_blank">https://docs.google.com/spreadsheets/d/1DpZUgJ3iEkKcFCNYv5PHz1bEcltX4UbQtB-uFTCc6S8/edit?usp=sharing</a>

### Homework 2: deadline December 15th

In the second homework we will analyze RNA-seq data. You will need to run a basic pipeline to analyze raw RNA-seq data and perform secondary analysis of given dataset using Deseq2.

Detailed description of the homework and scripts to help are located at https://drive.google.com/drive/folders/1C_awgIWc9UTXV0NwZ4IyT72FdtzKbmDz.

Datasets to analyze and biological questions (as well as you logins to the server) are available at https://docs.google.com/spreadsheets/d/1DpZUgJ3iEkKcFCNYv5PHz1bEcltX4UbQtB-uFTCc6S8/edit#gid=178367419.


### Homework 3: deadline December 19th

In this homework you will need to find fusion and viral transcripts in RNA-seq dataset.

https://docs.google.com/document/d/1lOHbz4aIzs8Tt_dNRuCDTs3Wxfg2L0Ax5ew9OcEsGfI/edit?usp=sharing

### Homework 4: deadline December 19th

For fourth homework you will need to chose one of recent scRNA-seq datasets and perform basic analysis:

* Preprocessing: filtering out bad cells and normalization
* UMAP + clustering
* Marker selection for clusters

I recommend using code snippets from the lecture or usual seurat tutorial at https://satijalab.org/seurat/v3.1/pbmc3k_tutorial.html


Dataset suggestions (you can analyze any dataset you want, these are just some suggestions):


* Transcriptome Analysis Reveals Non Foamy Rather Than Foamy Plaque Macrophages Are Proinflammatory in Atherosclerotic Murine Models, https://www.ahajournals.org/doi/pdf/10.1161/CIRCRESAHA.118.312804. Link to the dataset: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM3215435
* Suppressing miR-21 activity in tumor-associated macrophages promotes an antitumor immune response, https://www.jci.org/articles/view/127125. Link to the dataset: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE118931
* Coopted temporal patterning governs cellular hierarchy, heterogeneity and metabolism in Drosophila neuroblast tumors, https://elifesciences.org/articles/50375. Link to the dataset: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM3161971 
* Single-Cell RNA-Sequencing From Mouse Incisor Reveals Dental Epithelial Cell-Type Specific Genes https://www.frontiersin.org/articles/10.3389/fcell.2020.00841/full . Link to the dataset is  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM4407907 
* Cellular and molecular properties of neural progenitors in the developing mammalian hypothalamus, https://www.nature.com/articles/s41467-020-17890-2. Link to the dataset is https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM3330882 

You will pass this homework **by presenting results and comparing with the publication**.


