<!DOCTYPE html>
<html lang="en"><head>
<script src="intro_files/libs/clipboard/clipboard.min.js"></script>
<script src="intro_files/libs/quarto-html/tabby.min.js"></script>
<script src="intro_files/libs/quarto-html/popper.min.js"></script>
<script src="intro_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="intro_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="intro_files/libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="intro_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.0.37">

  <meta name="author" content="Konstantin Zaitsev">
  <title>Introduction to single-cell RNA-seq analysis</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="intro_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="intro_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    span.underline{text-decoration: underline;}
    div.column{display: inline-block; vertical-align: top; width: 50%;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="intro_files/libs/revealjs/dist/theme/quarto.css" id="theme">
  <link href="intro_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="intro_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="intro_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="intro_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-captioned.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-captioned) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-captioned.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-captioned .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-captioned .callout-caption  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-captioned.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-captioned.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-caption {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-caption {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-caption {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-captioned .callout-body > .callout-content > :last-child {
    margin-bottom: 0.5rem;
  }

  .callout.callout-captioned .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-captioned) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-caption {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-caption {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-caption {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-caption {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-caption {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="center">
  <h1 class="title">Introduction to single-cell RNA-seq analysis</h1>
  <p class="author">Konstantin Zaitsev</p>
  <p class="date">14/10/2022</p>
</section>

<section id="single-cell-rna-seq" class="slide level2">
<h2>Single cell RNA-seq</h2>
<div class="columns">
<div class="column" style="width:60%;">
<ul>
<li>RNA-seq is a snapshot of what is happening in the sample</li>
<li>Sample consists of many different cells and cell types</li>
<li>Single-cell RNA-seq - thousand of individual snapshots of many cells to capture the whole picture</li>
</ul>
</div><div class="column" style="width:40%;">
<p><img data-src="images/islets.png"></p>
</div>
</div>
</section>
<section id="why-single-cell-rna-seq" class="slide level2">
<h2>Why single-cell RNA-seq</h2>
<p>Heterogeneous populations:</p>
<ul>
<li>New cell subpopulations discovery</li>
<li>Comparison of similar cell subpopulations</li>
<li>Marker selection for cell subpopulations</li>
</ul>
<p>Homogeneous populations:</p>
<ul>
<li>Understanding heterogeneity</li>
<li>Cellular states and cellular processes</li>
</ul>
<p>Tracking of cell differentiation</p>
</section>
<section id="smart-seq2" class="slide level2">
<h2>Smart-seq2</h2>

<img data-src="images/smartseq2.png" class="r-stretch"><div class="footer">
<p>Paper: https://www.nature.com/articles/nprot.2014.006</p>
</div>
</section>
<section id="single-cell-rna-seq-of-myeloid-cells" class="slide level2">
<h2>Single-cell RNA-seq of myeloid cells</h2>
<div class="columns">
<div class="column" style="width:40%;">
<ul>
<li>Villani, Satija et al</li>
<li>Science, 2017</li>
<li>1152 cells</li>
</ul>
</div><div class="column" style="width:60%;">
<p><img data-src="images/myeloid_fix.png"></p>
</div>
</div>
<div class="footer">
<p>Paper: https://science.sciencemag.org/content/356/6335/eaah4573</p>
</div>
</section>
<section id="single-cell-rna-seq-of-myeloid-cells-1" class="slide level2">
<h2>Single-cell RNA-seq of myeloid cells</h2>

<img data-src="images/design_myeloid.png" class="r-stretch"><div class="footer">
<p>Paper: https://science.sciencemag.org/content/356/6335/eaah4573</p>
</div>
</section>
<section id="single-cell-rna-seq-of-melanoma" class="slide level2">
<h2>Single-cell RNA-seq of melanoma</h2>
<div class="columns">
<div class="column" style="width:30%;">
<ul>
<li>Tirosh, Izar et al</li>
<li>Science, 2016</li>
<li>4645 cells</li>
</ul>
</div><div class="column" style="width:70%;">
<p><img data-src="images/sc_cancer.png"></p>
</div>
</div>
<div class="footer">
<p>Paper: https://science.sciencemag.org/content/352/6282/189.long</p>
</div>
</section>
<section id="drop-seq-cell-2015" class="slide level2">
<h2>Drop-seq: Cell, 2015</h2>

<img data-src="images/dropseq.png" class="r-stretch"><div class="footer">
<p>Paper: https://www.cell.com/abstract/S0092-8674(15)00549-8</p>
</div>
</section>
<section id="x-chromium-machine-ncomms-2017" class="slide level2">
<h2>10x chromium machine: NComms, 2017</h2>

<img data-src="images/10x.png" class="r-stretch"><div class="footer">
<p>Paper: https://www.nature.com/articles/ncomms14049</p>
</div>
</section>
<section id="seq-well-nmeth-2017" class="slide level2">
<h2>Seq-Well: NMeth, 2017</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p><img data-src="images/seqwell_head.png"></p>
</div><div class="column" style="width:50%;">
<p><img data-src="images/seqwell_design.png"></p>
</div>
</div>
<div class="footer">
<p>Paper: https://www.nature.com/articles/nmeth.4179</p>
</div>
</section>
<section id="microwell-seq-cell-2018" class="slide level2">
<h2>Microwell-Seq: Cell, 2018</h2>

<img data-src="images/microwellseq.png" class="r-stretch"><div class="footer">
<p>Paper: https://www.ncbi.nlm.nih.gov/pubmed/29474909</p>
</div>
</section>
<section id="growth-of-single-cell-technologies" class="slide level2">
<h2>Growth of single-cell technologies</h2>

<img data-src="images/cells.png" class="r-stretch"><div class="footer">
<p>Human cell atlas white paper: https://arxiv.org/pdf/1810.05192.pdf</p>
</div>
</section>
<section id="growth-of-single-cell-technologies-1" class="slide level2">
<h2>Growth of single-cell technologies</h2>

<img data-src="images/papers.png" class="r-stretch"><div class="footer">
<p>Based on pubmed results</p>
</div>
</section>
<section id="main-challenges" class="slide level2">
<h2>Main challenges</h2>
<ul>
<li><strong>How to amplify and sequence small number of RNA (typical mammalian cell has only 200 000 mRNA molecules)?</strong></li>
<li>How to isolate cells?</li>
<li>How to work with big number of cells?</li>
</ul>
</section>
<section id="low-library-complexity" class="slide level2">
<h2>Low library complexity</h2>
<ul>
<li>Small numbers of mRNA molecules yield low complexity cDNA library</li>
<li>cDNA molecules to be amplified by PCR</li>
<li>We don’t want to sequence tons of PCR duplicates</li>
</ul>

<img data-src="images/pcr.png" class="r-stretch"></section>
<section id="template-switching-pcr" class="slide level2">
<h2>Template-switching PCR</h2>
<div class="columns">
<div class="column" style="width:50%;">
<ul>
<li>Switching Mechanism At the 5’ end of RNA Template (SMART)</li>
<li>Robust for low input libraries</li>
</ul>
</div><div class="column" style="width:50%;">
<p><img data-src="images/template_switch.png"></p>
</div>
</div>
</section>
<section id="drop-seq-schematics" class="slide level2">
<h2>Drop-seq schematics</h2>

<img data-src="images/scheme_1.png" class="r-stretch"><div class="footer">
<p>Paper: https://www.cell.com/abstract/S0092-8674(15)00549-8</p>
</div>
</section>
<section id="drop-seq-microfluidics" class="slide level2">
<h2>Drop-seq microfluidics</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p><img data-src="images/microfluidics1.jpg"></p>
</div><div class="column" style="width:50%;">
<p><img data-src="images/microfluidics2.jpg.png"></p>
</div>
</div>
<div class="footer">
<p>Paper: https://www.cell.com/abstract/S0092-8674(15)00549-8</p>
</div>
</section>
<section id="drop-seq-microfluidics-1" class="slide level2">
<h2>Drop-seq microfluidics</h2>

<img data-src="images/dropseq_gif.gif" class="r-stretch"></section>
<section id="drop-seq-schematics-1" class="slide level2">
<h2>Drop-seq schematics</h2>

<img data-src="images/scheme_2.png" class="r-stretch"><div class="footer">
<p>Paper: https://www.cell.com/abstract/S0092-8674(15)00549-8</p>
</div>
</section>
<section id="barcoding" class="slide level2">
<h2>Barcoding</h2>
<div class="r-stack">
<p><img data-src="images/barcoded_beads-01.png" class="fragment fade-out" data-fragment-index="0"></p>
<p><img data-src="images/barcoded_beads-03.png" class="fragment currently-visible" data-fragment-index="0"></p>
<p><img data-src="images/barcoded_beads-02.png" class="fragment"></p>
<p><img data-src="images/barcoded_beads-04.png" class="fragment"></p>
</div>
</section>
<section id="drop-seq-schematics-2" class="slide level2">
<h2>Drop-seq schematics</h2>

<img data-src="images/scheme_3.png" class="r-stretch"><div class="footer">
<p>Paper: https://www.cell.com/abstract/S0092-8674(15)00549-8</p>
</div>
</section>
<section id="sequencing" class="slide level2">
<h2>Sequencing</h2>
<div class="r-stack">
<p><img data-src="images/umis1.png" class="fragment fade-out" data-fragment-index="0"></p>
<p><img data-src="images/umis2.png" class="fragment currently-visible" data-fragment-index="0"></p>
<p><img data-src="images/umis3.png" class="fragment"></p>
<p><img data-src="images/umis4.png" class="fragment"></p>
<p><img data-src="images/umis5.png" class="fragment"></p>
<p><img data-src="images/umis6.png" class="fragment"></p>
<p><img data-src="images/umis7.png" class="fragment"></p>
<p><img data-src="images/umis8.png" class="fragment"></p>
</div>
</section>
<section id="all-questions-were-addressed" class="slide level2">
<h2>All questions were addressed</h2>
<ul>
<li>Low input mRNA – template switching PCR</li>
<li>Cell isolation – microfluidics</li>
<li>Read identification – Cell barcodes</li>
<li>Dealing with PCR duplicates – Cell/UMI barcodes</li>
</ul>
</section>
<section id="all-questions-were-addressed-1" class="slide level2">
<h2>All questions were addressed</h2>
<ul>
<li>Low input mRNA – template switching PCR</li>
<li><strong>Cell isolation – microfluidics (might vary from technology to technology)</strong></li>
<li><strong>Barcoding with beads proved to be very effective</strong></li>
</ul>
</section>
<section id="dataset-for-today" class="slide level2">
<h2>Dataset for today:</h2>
<ul>
<li><a href="https://www.10xgenomics.com/resources/datasets/1-k-pbm-cs-from-a-healthy-donor-v-3-chemistry-3-standard-3-0-0" target="_blank">https://www.10xgenomics.com/resources/datasets/1-k-pbm-cs-from-a-healthy-donor-v-3-chemistry-3-standard-3-0-0</a></li>
</ul>

<img data-src="images/1k_v3.png" class="r-stretch"></section>
<section id="dataset-summary" class="slide level2">
<h2>Dataset summary</h2>
<ul>
<li><a href="https://cf.10xgenomics.com/samples/cell-exp/3.0.0/pbmc_1k_v3/pbmc_1k_v3_web_summary.html" target="_blank">https://cf.10xgenomics.com/samples/cell-exp/3.0.0/pbmc_1k_v3/pbmc_1k_v3_web_summary.html</a></li>
</ul>
</section>
<section id="what-is-noise" class="slide level2">
<h2>What is noise?</h2>

<img data-src="images/filtering_procedure.png" class="r-stretch"></section>
<section id="what-is-noise-1" class="slide level2">
<h2>What is noise?</h2>
<p>There is cell-free RNA in the cellular suspension that will be captured in empty droplets (with beads), and we must distinguish cells from empty droplets:</p>
<ul>
<li>We can set a hard threshold on UMI to filter noise from actual cells</li>
<li>We can calculate noise signature and test every cell against this signature (like <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1662-y">emptyDrops</a>)</li>
</ul>
</section>
<section id="our-setup" class="slide level2">
<h2>Our setup</h2>
<ul>
<li>Go to <a href="https://drive.google.com/drive/folders/18J9NgzLSLnBjmz_YQsMdOSKKK-lWhSVw?usp=sharing" class="uri">https://drive.google.com/drive/folders/18J9NgzLSLnBjmz_YQsMdOSKKK-lWhSVw?usp=sharing</a></li>
<li>Download <code>materials.zip</code> from the google folder</li>
<li>Create new RStudio project</li>
<li>Upload the script to your r-studio session</li>
<li>Open file <code>scrna-seq-intro.R</code></li>
</ul>
</section>
<section id="our-setup-create-project" class="slide level2">
<h2>Our setup: create project</h2>
<ul>
<li>Create new project</li>
<li>New Directory -&gt; New Project -&gt; chose a name and folder</li>
</ul>
<div class="columns">
<div class="column" style="width:50%;">
<p><img data-src="images/new_project.png"></p>
</div><div class="column" style="width:50%;">
<p><img data-src="images/new_project_wizard.png"></p>
</div>
</div>
</section>
<section id="our-setup-upload-materials" class="slide level2">
<h2>Our setup: upload materials</h2>
<ul>
<li>Upload the materials zip</li>
</ul>
<div class="columns">
<div class="column" style="width:50%;">
<p><img data-src="images/upload_button.png"></p>
</div><div class="column" style="width:50%;">
<p><img data-src="images/select_materials.png"></p>
</div>
</div>
</section>
<section id="our-setup-final-files" class="slide level2">
<h2>Our setup: final files</h2>

<img data-src="images/files.png" class="r-stretch"></section>
<section id="our-setup-1" class="slide level2">
<h2>Our setup</h2>
<ul>
<li>You will see chunks of code appearing on my slides</li>
<li>Follow along!</li>
<li>Select the code and press Ctrl + Enter to run the chunk</li>
</ul>
</section>
<section id="basic-steps-to-analysis-of-scrna-seq" class="slide level2">
<h2>Basic steps to analysis of scRNA-seq</h2>
<ul>
<li><strong>Filtering out “bad” cells</strong></li>
<li>Normalizing expression levels: (scaling and log2 normalizing)</li>
<li>PCA</li>
<li>Visualization (tSNE or UMAP)</li>
<li>Clustering</li>
<li>Cellular subset annotation</li>
</ul>
</section>
<section id="loading-all-the-librarires" class="slide level2">
<h2>Loading all the librarires</h2>
<p>We will use Seurat - R toolkit for single-cell genomics</p>
<p><a href="https://satijalab.org/seurat/" target="_blank">https://satijalab.org/seurat/</a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="fu">library</span>(Matrix)</span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="fu">library</span>(MAST)</span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="fu">library</span>(dplyr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">

</div>
</section>
<section id="loading-the-data" class="slide level2">
<h2>Loading the data</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>data <span class="ot">&lt;-</span> <span class="fu">Read10X</span>(<span class="st">"filtered_feature_bc_matrix/"</span>)</span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="fu">dim</span>(data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 33538  1222</code></pre>
</div>
</div>
<ul>
<li>33538 genes</li>
<li>1222 cells</li>
</ul>
</section>
<section id="umi-distribution" class="slide level2">
<h2>UMI distribution</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>plotData <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb4-2"><a href="#cb4-2"></a>  umis <span class="ot">&lt;-</span> <span class="fu">colSums</span>(data)</span>
<span id="cb4-3"><a href="#cb4-3"></a>)</span>
<span id="cb4-4"><a href="#cb4-4"></a><span class="fu">ggplot</span>(<span class="at">data=</span>plotData, <span class="fu">aes</span>(<span class="at">x=</span>umis)) <span class="sc">+</span></span>
<span id="cb4-5"><a href="#cb4-5"></a>  <span class="fu">geom_histogram</span>() <span class="sc">+</span> <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell fragment">
<div class="cell-output-display">
<p><img data-src="intro_files/figure-revealjs/unnamed-chunk-8-1.png" width="960"></p>
</div>
</div>
</section>
<section id="filtering-genes-and-barcodes" class="slide level2">
<h2>Filtering (genes and barcodes)</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>seurat <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(data, <span class="at">min.cells =</span> <span class="dv">10</span>, <span class="at">min.features =</span> <span class="dv">10</span>)</span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="fu">dim</span>(seurat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 12566  1222</code></pre>
</div>
</div>
<ul>
<li>Only 12566 genes left after filtering (detected in <span class="math inline">\(\leq 10\)</span> cells)</li>
</ul>
</section>
<section id="filtering-genes-and-barcodes-1" class="slide level2">
<h2>Filtering (genes and barcodes)</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="fu">FeatureScatter</span>(seurat, <span class="st">"nCount_RNA"</span>, <span class="st">"nFeature_RNA"</span>) <span class="sc">+</span> </span>
<span id="cb7-2"><a href="#cb7-2"></a>  <span class="fu">scale_x_log10</span>() <span class="sc">+</span> </span>
<span id="cb7-3"><a href="#cb7-3"></a>  <span class="fu">scale_y_log10</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="intro_files/figure-revealjs/unnamed-chunk-12-1.png" width="960" class="r-stretch"></section>
<section id="mitochondrial-content" class="slide level2">
<h2>Mitochondrial content</h2>
<ul>
<li>High numbers of mitochondrial transcripts are associated with apoptotic<sup>1</sup>, stressed<sup>2</sup> and low-quality cells in general</li>
<li>mtDNA% threshold depends highly on the tissue type and the questions being investigated<sup>3</sup></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a>seurat[[<span class="st">"percent.mt"</span>]] <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(seurat, <span class="at">pattern =</span> <span class="st">"^MT-"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<aside><ol class="aside-footnotes"><li id="fn1" role="doc-endnote"><p>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8599307/</p></li><li id="fn2" role="doc-endnote"><p>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC126185/</p></li><li id="fn3" role="doc-endnote"><p>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6072887/</p></li></ol></aside></section>
<section id="filtering-genes-and-barcodes-2" class="slide level2">
<h2>Filtering (genes and barcodes)</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="fu">FeatureScatter</span>(seurat, <span class="st">"nCount_RNA"</span>, <span class="st">"percent.mt"</span>) <span class="sc">+</span> </span>
<span id="cb9-2"><a href="#cb9-2"></a>  <span class="fu">scale_x_log10</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="intro_files/figure-revealjs/unnamed-chunk-16-1.png" width="960" class="r-stretch"></section>
<section id="filtering-genes-and-barcodes-3" class="slide level2">
<h2>Filtering (genes and barcodes)</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a><span class="fu">FeatureScatter</span>(seurat, <span class="st">"nFeature_RNA"</span>, <span class="st">"percent.mt"</span>) <span class="sc">+</span> </span>
<span id="cb10-2"><a href="#cb10-2"></a>  <span class="fu">scale_x_log10</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="intro_files/figure-revealjs/unnamed-chunk-18-1.png" width="960" class="r-stretch"></section>
<section id="filtering-genes-and-barcodes-4" class="slide level2">
<h2>Filtering (genes and barcodes)</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="fu">FeatureScatter</span>(seurat, <span class="st">"nFeature_RNA"</span>, <span class="st">"percent.mt"</span>) <span class="sc">+</span> <span class="fu">scale_x_log10</span>() <span class="sc">+</span></span>
<span id="cb11-2"><a href="#cb11-2"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">15</span>, <span class="at">lty=</span><span class="dv">2</span>) <span class="sc">+</span> <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">1200</span>, <span class="at">lty=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="intro_files/figure-revealjs/unnamed-chunk-20-1.png" width="960" class="r-stretch"></section>
<section id="filtering-genes-and-barcodes-5" class="slide level2">
<h2>Filtering (genes and barcodes)</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a>seurat <span class="ot">&lt;-</span> <span class="fu">subset</span>(seurat, <span class="at">subset =</span> nFeature_RNA <span class="sc">&gt;</span> <span class="dv">1200</span> <span class="sc">&amp;</span> percent.mt <span class="sc">&lt;</span> <span class="dv">15</span>)</span>
<span id="cb12-2"><a href="#cb12-2"></a><span class="fu">dim</span>(seurat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 12566  1011</code></pre>
</div>
</div>
<ul>
<li>we filtered out 211 droplets due to low quality</li>
</ul>
</section>
<section id="basic-steps-to-analysis-of-scrna-seq-1" class="slide level2">
<h2>Basic steps to analysis of scRNA-seq</h2>
<ul>
<li>Filtering out “bad” cells</li>
<li><strong>Normalizing expression levels: (scaling and log2 normalizing)</strong></li>
<li>PCA</li>
<li>Visualization (tSNE or UMAP)</li>
<li>Clustering</li>
<li>Cellular subset annotation</li>
</ul>
</section>
<section id="normalization-old-way" class="slide level2">
<h2>Normalization (old way)</h2>
<ul>
<li>Scaling expression to 10 000 UMIs (instead of million in RPM)</li>
<li>Finding features with high variance (features above mean/variance trend)</li>
<li>Scaling data (for PCA and so on)</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a>seurat <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(seurat, <span class="at">normalization.method =</span> <span class="st">"LogNormalize"</span>, <span class="at">scale.factor =</span> <span class="dv">10000</span>)</span>
<span id="cb14-2"><a href="#cb14-2"></a>seurat <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(seurat, <span class="at">selection.method =</span> <span class="st">"vst"</span>, <span class="at">nfeatures =</span> <span class="dv">2000</span>)</span>
<span id="cb14-3"><a href="#cb14-3"></a>seurat <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(seurat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="normalization---sctransform" class="slide level2">
<h2>Normalization - SCTransform</h2>
<div class="cell" data-hash="intro_cache/revealjs/unnamed-chunk-26_b51cd0f2540024ab8fdc58dcb7908ac6">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a>seurat <span class="ot">&lt;-</span> <span class="fu">SCTransform</span>(seurat, <span class="at">vars.to.regress =</span> <span class="st">"percent.mt"</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>

<img data-src="images/sctransform.png" class="r-stretch"><div class="footer">
<p>Paper: https://www.biorxiv.org/content/biorxiv/early/2019/03/18/576827.full.pdf</p>
</div>
</section>
<section id="variable-genes" class="slide level2">
<h2>Variable genes</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a><span class="fu">VariableFeaturePlot</span>(seurat) <span class="sc">+</span> <span class="fu">scale_y_log10</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="intro_files/figure-revealjs/unnamed-chunk-28-1.png" width="960" class="r-stretch"></section>
<section id="variable-genes-1" class="slide level2">
<h2>Variable genes</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a>top10_variable_genes <span class="ot">&lt;-</span> <span class="fu">head</span>(<span class="fu">VariableFeatures</span>(seurat), <span class="dv">10</span>)</span>
<span id="cb17-2"><a href="#cb17-2"></a><span class="fu">VariableFeaturePlot</span>(seurat) <span class="sc">%&gt;%</span> </span>
<span id="cb17-3"><a href="#cb17-3"></a>  <span class="fu">LabelPoints</span>(<span class="at">points =</span> top10_variable_genes, <span class="at">repel =</span> <span class="cn">TRUE</span>) <span class="sc">+</span></span>
<span id="cb17-4"><a href="#cb17-4"></a>  <span class="fu">scale_y_log10</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="intro_files/figure-revealjs/unnamed-chunk-30-1.png" width="960" class="r-stretch"></section>
<section id="basic-steps-to-analysis-of-scrna-seq-2" class="slide level2">
<h2>Basic steps to analysis of scRNA-seq</h2>
<ul>
<li>Filtering out “bad” barcodes</li>
<li>Normalizing expression levels: (scaling and log2 normalizing)</li>
<li><strong>PCA</strong></li>
<li>Visualization (tSNE or UMAP)</li>
<li>Clustering</li>
<li>Cellular subset annotation</li>
</ul>
</section>
<section id="high-dimensionality-of-scrna-seq" class="slide level2">
<h2>High-dimensionality of scRNA-seq</h2>
<p>Initially matrix is very large in size, this causes different kind of issues:</p>
<ul>
<li>Some algorithms are just slow when performed in this high-dimensionality data</li>
<li>Curse of dimensionality</li>
</ul>
<p>We usually take several steps to reduce dims before creating 2d clustered map of our dataset</p>
<ul>
<li>Keeping only variable genes (since those introduce variance to the dataset)</li>
<li>PCA will reduce dimensionality to 20-30 first components</li>
</ul>
</section>
<section id="pca" class="slide level2">
<h2>PCA</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a>seurat <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(seurat, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb18-2"><a href="#cb18-2"></a><span class="fu">ElbowPlot</span>(seurat, <span class="at">ndims =</span> <span class="dv">50</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="intro_files/figure-revealjs/unnamed-chunk-32-1.png" width="960" class="r-stretch"></section>
<section id="basic-steps-to-analysis-of-scrna-seq-3" class="slide level2">
<h2>Basic steps to analysis of scRNA-seq</h2>
<ul>
<li>Filtering out “bad” barcodes</li>
<li>Normalizing expression levels: (scaling and log2 normalizing)</li>
<li>PCA</li>
<li><strong>Visualization (tSNE or UMAP)</strong></li>
<li>Clustering</li>
<li>Cellular subset annotation</li>
</ul>
</section>
<section id="visualization" class="slide level2">
<h2>Visualization</h2>
<p>Both tSNE and UMAP will put our data after PCA into a 2D plane:</p>
<ul>
<li>Dots that are close to each other are cells that are transcriptionally similar to each other</li>
<li>Dots that are far from each other are cells that are trasncriptionally different from each other</li>
</ul>
<p>Lets look at dimensionality of each cell:</p>
<ul>
<li>Variable genes only: 30k+ -&gt; 2-3k of variable genes</li>
<li>PCA: 2-3k of variable genes -&gt; 20-30 principal components</li>
<li>tSNE or UMAP: 20-30 PCs -&gt; 2d or 3d plots</li>
</ul>
</section>
<section id="tsne" class="slide level2">
<h2>tSNE</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a>seurat <span class="ot">&lt;-</span> <span class="fu">RunTSNE</span>(seurat, <span class="at">dims=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>)</span>
<span id="cb19-2"><a href="#cb19-2"></a><span class="fu">DimPlot</span>(seurat, <span class="at">reduction =</span> <span class="st">"tsne"</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="intro_files/figure-revealjs/unnamed-chunk-34-1.png" width="960" class="r-stretch"><div class="footer">
<p>Paper: http://www.jmlr.org/papers/volume9/vandermaaten08a/vandermaaten08a.pdf</p>
</div>
</section>
<section id="tsne-1" class="slide level2">
<h2>tSNE</h2>
<p>tSNE - “t-Distributed Stochastic Neighbor Embedding (t-SNE) is a technique for dimensionality reduction that is particularly well suited for the visualization of high-dimensional datasets”</p>
<ul>
<li>We can calculate distances in original space, and then calculate conditional probabilities <span class="math inline">\(p_{i| j}\)</span> that point <span class="math inline">\(i\)</span> would choose point <span class="math inline">\(j\)</span> as a neighbor. <span class="math inline">\(p_{i| j}\)</span> are proportional to “distances” from <span class="math inline">\(i\)</span> to all other points (actually probability density around point <span class="math inline">\(i\)</span>, but it doesn’t matter here).</li>
<li>Once all <span class="math inline">\(p_{i, j}\)</span> are calculated in original space we try to find such 2d/3d space that would have similar probabilities</li>
</ul>
<div class="footer">
<p>Paper: http://www.jmlr.org/papers/volume9/vandermaaten08a/vandermaaten08a.pdf</p>
</div>
</section>
<section id="umap" class="slide level2">
<h2>UMAP</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a>seurat <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(seurat, <span class="at">dims=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>)</span>
<span id="cb20-2"><a href="#cb20-2"></a><span class="fu">DimPlot</span>(seurat, <span class="at">reduction =</span> <span class="st">"umap"</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="intro_files/figure-revealjs/unnamed-chunk-36-1.png" width="960" class="r-stretch"><div class="footer">
<p>Paper: https://arxiv.org/pdf/1802.03426.pdf</p>
</div>
</section>
<section id="umap-1" class="slide level2">
<h2>UMAP</h2>

<img data-src="images/umap.png" class="r-stretch"><div class="footer">
<p>Paper: https://arxiv.org/pdf/1802.03426.pdf</p>
</div>
</section>
<section id="umap-2" class="slide level2">
<h2>UMAP</h2>
<p><a href="https://pair-code.github.io/understanding-umap/" class="uri">https://pair-code.github.io/understanding-umap/</a></p>
</section>
<section id="comparing-two" class="slide level2">
<h2>Comparing two</h2>
<div class="r-stack">
<div class="fragment fade-out" data-fragment-index="0">
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a><span class="fu">DimPlot</span>(seurat, <span class="at">reduction =</span> <span class="st">"tsne"</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img data-src="intro_files/figure-revealjs/unnamed-chunk-38-1.png" width="960"></p>
</div>
</div>
</div>
<div class="fragment currently-visible" data-fragment-index="0">
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a><span class="fu">DimPlot</span>(seurat, <span class="at">reduction =</span> <span class="st">"umap"</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img data-src="intro_files/figure-revealjs/unnamed-chunk-40-1.png" width="960"></p>
</div>
</div>
</div>
</div>
</section>
<section id="basic-steps-to-analysis-of-scrna-seq-4" class="slide level2">
<h2>Basic steps to analysis of scRNA-seq</h2>
<ul>
<li>Filtering out “bad” barcodes</li>
<li>Normalizing expression levels: (scaling and log2 normalizing)</li>
<li>PCA</li>
<li>Visualization (tSNE or UMAP)</li>
<li><strong>Clustering</strong></li>
<li>Cellular subset annotation</li>
</ul>
</section>
<section id="clustering-and-annotation" class="slide level2">
<h2>Clustering and annotation</h2>
<p>Clustering:</p>
<ul>
<li>Graph-based clustering (preferred)</li>
<li>K-means</li>
</ul>
<p>Annotation:</p>
<ul>
<li>First, check known markers</li>
<li>For each cluster, perform differential expression: cluster against all others</li>
<li>Top DE genes expected to be highly distinctive marker genes</li>
</ul>
</section>
<section id="clustering" class="slide level2">
<h2>Clustering</h2>
<div class="columns">
<div class="column" style="width:50%;">
<ul>
<li>Instead of defining clusters based on distance we first find “shared nearest neighbors”</li>
<li>Cells that have a lot of neighbors in common, most likely “live in the same neighborhood”</li>
<li>Algorithm is trying to find such neighborhoods</li>
</ul>
</div><div class="column" style="width:50%;">
<p><img data-src="images/snn-cliq.png"></p>
</div>
</div>
<div class="footer">
<p>Paper: https://academic.oup.com/bioinformatics/article/31/12/1974/214505</p>
</div>
</section>
<section id="clustering-1" class="slide level2">
<h2>Clustering</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a>seurat <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(seurat, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb23-2"><a href="#cb23-2"></a>seurat <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(seurat, <span class="at">resolution=</span><span class="fl">0.6</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="clustering-2" class="slide level2">
<h2>Clustering</h2>
<div class="r-stack">
<div class="fragment fade-out" data-fragment-index="0">
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a><span class="fu">DimPlot</span>(seurat, <span class="at">reduction =</span> <span class="st">"tsne"</span>, <span class="at">label =</span> <span class="cn">TRUE</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img data-src="intro_files/figure-revealjs/unnamed-chunk-44-1.png" width="960"></p>
</div>
</div>
</div>
<div class="fragment currently-visible" data-fragment-index="0">
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a><span class="fu">DimPlot</span>(seurat, <span class="at">reduction =</span> <span class="st">"umap"</span>, <span class="at">label =</span> <span class="cn">TRUE</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img data-src="intro_files/figure-revealjs/unnamed-chunk-46-1.png" width="960"></p>
</div>
</div>
</div>
</div>
</section>
<section id="pbmc" class="slide level2">
<h2>PBMC</h2>

<img data-src="images/pbmc.png" class="r-stretch"></section>
<section id="annotation" class="slide level2">
<h2>Annotation</h2>
<ul>
<li>Known markers: CD14, CD79A, CD3D are known markers of Monocytes, B cells and T cells respectively</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a><span class="fu">FeaturePlot</span>(seurat, <span class="fu">c</span>(<span class="st">"CD14"</span>, <span class="st">"CD79A"</span>, <span class="st">"CD3D"</span>), <span class="at">cols=</span><span class="fu">c</span>(<span class="st">"grey"</span>, <span class="st">"red"</span>), <span class="at">reduction=</span><span class="st">"umap"</span>, <span class="at">ncol=</span><span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="intro_files/figure-revealjs/unnamed-chunk-48-1.png" width="960" class="r-stretch"></section>
<section id="annotation-1" class="slide level2">
<h2>Annotation</h2>
<p>We can run differential expression test to identify markers automatically using MAST test</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a>allMarkers <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(seurat, <span class="at">max.cells.per.ident =</span> <span class="dv">100</span>, <span class="at">test.use =</span> <span class="st">"MAST"</span>, <span class="at">only.pos =</span> T)</span>
<span id="cb27-2"><a href="#cb27-2"></a>goodMarkers <span class="ot">&lt;-</span> allMarkers <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(cluster) <span class="sc">%&gt;%</span> <span class="fu">top_n</span>(<span class="at">n =</span> <span class="dv">1</span>, <span class="at">wt =</span> avg_log2FC) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(gene)</span>
<span id="cb27-3"><a href="#cb27-3"></a>goodMarkers</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "S100A8" "IL7R"   "NOSIP"  "CCL5"   "IGHM"   "CST3"   "IGHA1"  "GNLY"  
[9] "CD8B"  </code></pre>
</div>
</div>
</section>
<section id="ways-to-show-expression" class="slide level2">
<h2>Ways to show expression</h2>
<p>We usually either show expression <strong>on top of reduction plot</strong>, or show a violin plot for expression</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1"></a><span class="fu">FeaturePlot</span>(seurat, goodMarkers[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>], <span class="at">cols=</span><span class="fu">c</span>(<span class="st">"grey"</span>, <span class="st">"red"</span>), <span class="at">reduction=</span><span class="st">"umap"</span>, <span class="at">ncol=</span><span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="intro_files/figure-revealjs/unnamed-chunk-52-1.png" width="960" class="r-stretch"></section>
<section id="ways-to-show-expression-1" class="slide level2">
<h2>Ways to show expression</h2>
<p>We usually either show expression of top of reduction plot, or show a <strong>violin plot</strong> for expression</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1"></a><span class="fu">VlnPlot</span>(seurat, goodMarkers[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>], <span class="at">pt.size =</span> <span class="fl">0.1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="intro_files/figure-revealjs/unnamed-chunk-54-1.png" width="960" class="r-stretch"></section>
<section id="communication-is-important" class="slide level2">
<h2>Communication is important</h2>
<ul>
<li>When clusters are found we want to identify which cell subsets are presented, to “annotate” them</li>
<li>If you are a bioinformatician and you are a single-cell RNA-seq dataset that have been designed/done by you, this is a perfect to time to go and talk to a biologist who performed/designed the experiment</li>
<li>If you are a biologist who designed/performed single-cell RNA-seq experiment, chances are, you know all cellular subsets and markers better than almost anyone else</li>
<li>This is where you communicate and try to make sense of the data</li>
</ul>
</section>
<section id="differential-expression-in-single-cell-rna-seq" class="slide level2">
<h2>Differential expression in single-cell RNA-seq</h2>
<ul>
<li>Within one dataset (comparing clusters, comparing a cluster vs all other clusters)</li>
<li>Marker identification, population annotation</li>
<li>This is OK, statistical power is enough to make conclusions about cellular populations within single-dataset</li>
</ul>
</section>
<section id="differential-expression-in-single-cell-rna-seq-1" class="slide level2">
<h2>Differential expression in single-cell RNA-seq</h2>
<ul>
<li>Between two datasets (comparing the same cells between samples A vs B)</li>
<li>Only one biological replicate is present in the data :(</li>
<li>Different noise levels in the samples</li>
</ul>
</section>
<section id="differential-expression-in-single-cell-rna-seq-2" class="slide level2">
<h2>Differential expression in single-cell RNA-seq</h2>
<ul>
<li>Between groups of samples (comparing the same cells between several samples of condition A vs several samples of condition B)</li>
<li>More biological replicates :)</li>
<li>Still different noise levels within the samples and within the groups</li>
<li>Can we do it correctly?</li>
</ul>
</section>
<section id="differential-expression-in-single-cell-rna-seq-3" class="slide level2">
<h2>Differential expression in single-cell RNA-seq</h2>

<img data-src="images/de_false_discoveries.png" class="r-stretch"><div class="footer">
<p>Paper:https://www.nature.com/articles/s41467-021-25960-2.pdf</p>
</div>
</section>
<section id="differential-expression-in-single-cell-rna-seq-4" class="slide level2">
<h2>Differential expression in single-cell RNA-seq</h2>

<img data-src="images/de_test_design.png" class="r-stretch"><div class="footer">
<p>Paper:https://www.nature.com/articles/s41467-021-25960-2.pdf</p>
</div>
</section>
<section id="differential-expression-in-single-cell-rna-seq-5" class="slide level2">
<h2>Differential expression in single-cell RNA-seq</h2>

<img data-src="images/de_findings.png" class="r-stretch"><div class="footer">
<p>Paper:https://www.nature.com/articles/s41467-021-25960-2.pdf</p>
</div>
</section>
<section id="takeaway-messages-about-differential-expression" class="slide level2">
<h2>Takeaway messages about differential expression</h2>
<ul>
<li>When you have an option - make more biological replicates and stick to pseudo-bulk DE. This will reduce false discoveries</li>
<li>Latest chemistry from 10x enables samples multiplexing easier than before <a href="https://www.10xgenomics.com/products/single-cell-fixed-rna-profiling" class="uri">https://www.10xgenomics.com/products/single-cell-fixed-rna-profiling</a></li>
</ul>
</section>
<section id="questions" class="slide level2">
<h2>Questions?</h2>

<img src="libs/itmo_small_white_eng.png" class="slide-logo r-stretch"><div class="footer footer-default">
<p>INC RAS - 2022</p>
</div>
</section>

    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="intro_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="intro_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="intro_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="intro_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="intro_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="intro_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="intro_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="intro_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="intro_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="intro_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'smaller': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        function fireSlideChanged(previousSlide, currentSlide) {

          // dispatch for htmlwidgets
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for reveal
        if (window.Reveal) {
          window.Reveal.addEventListener("slidechanged", function(event) {
            fireSlideChanged(event.previousSlide, event.currentSlide);
          });
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        target: function(trigger) {
          return trigger.previousElementSibling;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        setTimeout(function() {
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn) {
        const config = {
          allowHTML: true,
          content: contentFn,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'quarto-reveal',
          placement: 'bottom-start'
        };
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          let href = ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const cites = ref.parentNode.getAttribute('data-cites').split(' ');
        tippyHover(ref, function() {
          var popup = window.document.createElement('div');
          cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    });
    </script>
    

</body></html>