---
title: "Gene expression 2019: syllabus"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Before we start

In this course, we (Konstantin Zaitsev and Alexander Tkachenko) will try to explain most of the essential things you need to now about gene expression. We will go from explaining ways to capture RNA and measure RNA abundance to discussing how to properly design gene expression study and what the best practices are.

Authors of this course:

1. Konstantin Zaitsev, https://www.researchgate.net/profile/Konstantin_Zaitsev, https://stepik.org/course/512/
2. Alexander Tkachenko, https://publons.com/researcher/3041889/alexander-tkachenko/, https://stepik.org/course/94/


Most of the material updates will be published here.

Disclaimers:

1. All the course materials will be in English
2. By "gene expression" we almost always mean "RNA expression" if not said otherwise

## Syllabus

+ Central dogma of molecular biology, structure of gene, types of RNA, structure of RNA, transcription, reverse transcription, qPCR, FISH, <a href="https://konsolerr.github.io/gene_expression_2019/facs" target="_blank">FACS</a>
+ Microarray: quantification, normalization, basic analysis
+ RNA-seq: alignment, quantification, QC, normalization, basic analysis
+ Overall quality control: PCA, clustering, outlier detection
+ Overall quality control: batch correction
+ Differential expression (DE): limma for microarray, Deseq2 for RNA-seq
+ Downstream analysis: pathway/gene set enrichment analysis
+ Downstream analysis: gene expression deconvolution
+ Transcriptome assembly, functional annotation
+ Single-cell transcriptomics: Seurat basic analysis
+ Single-cell transcriptomics: Trajectory analysis, RNA velocity, optimal transport
+ Visual data exploration: phantasus, JBR genome browser
+ Experimental design of gene expression study

## Prerequisites

Most of the lessons will be in format of seminars: lectures and practical part with your laptops. Prior the start of the course, you are required to install latest version of R (https://cran.r-project.org/bin/) and Rstudio (https://www.rstudio.com/) to your laptop. Most of the practices we will be working in R and only for several we will require a command line.

## Homeworks 

There will be 4 projects homeworks during this course:

1. Microarray dataset: processing, QC, analysis, DE
2. RNA-seq dataset: processing (from fastq), QC, analysis, DE, downstream analysis (pathways)
3. Transcriptome assembly and functional annotation
4. Single-cell RNA-seq dataset: analysis with Seurat


## Below we will put links to slides and other useful links

### Lecture materials

1. Lecture 1: <a href="https://konsolerr.github.io/gene_expression_2019/intro" target="_blank">introduction</a> and <a href="https://konsolerr.github.io/gene_expression_2019/facs" target="_blank">FACS</a>
2. Lecture 2: <a href="https://konsolerr.github.io/gene_expression_2019/transcription/transcription.pdf" target="_blank">transcription</a>
3. Lecture 3: Microarrays <a href="https://konsolerr.github.io/gene_expression_2019/microarray/microarray.html" target="_blank">part 1</a> and <a href="https://konsolerr.github.io/gene_expression_2019/microarray/microarray2.html" target="_blank">part 2</a>
4. Lecture 4: Secondary analysis <a href="https://konsolerr.github.io/gene_expression_2019/secondary_analysis" target="_blank">slides</a> and <a href="https://konsolerr.github.io/gene_expression_2019/secondary_analysis/secondary_analysis.zip" target="_blank">materials</a>
5. Lecture 5: RNA-seq <a href="https://konsolerr.github.io/gene_expression_2019/rna-seq" target="_blank">slides</a> and <a href="https://konsolerr.github.io/gene_expression_2019/rna-seq/rna_seq_materials.zip" target="_blank">materials</a>
6. Lecture 6: Transcriptome assembly and annotation. <a href="https://konsolerr.github.io/gene_expression_2019/assembly/Transcriptome_assembly.pdf" target="_blank">Slides</a>
7. Lecture 7: gene expression deconvolution <a href="https://konsolerr.github.io/gene_expression_2019/deconvolution" target="_blank">slides</a> and <a href="https://konsolerr.github.io/gene_expression_2019/deconvolution/deconvolution.R" target="_blank">code</a>.

### Homework 1 

The first homework is to analyze a microarray gene expression dataset. All the datasets will be available by GEO id, so you can find them at GEO omnibus and also use `GEOquery::getGEO("GSEXXXXX")` to get the expression data into R session.

Basically you needed to perform basic steps for a dataset: normalize and make sense of the PCA plot. Then you need to disgn differential exression to address the general question for your homework. Once you perform DE I want you to highlight genes that are different and perform pathway enrichment.

Datasets IDs are listed in the table below:

https://docs.google.com/spreadsheets/d/15tEZ3fADDdxzIxdBfFhAozQUMb-chAjX14qtdRJKtP8/edit?usp=sharing

