---
title: "Gene expression 2019: syllabus"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Before we start

In this course, we (Konstantin Zaitsev and Alexander Tkachenko) will try to explain most of the essential things you need to now about gene expression. We will go from explaining ways to capture RNA and measure RNA abundance to discussing how to properly design gene expression study and what the best practices are.

Authors of this course:

1. Konstantin Zaitsev, https://www.researchgate.net/profile/Konstantin_Zaitsev, https://stepik.org/course/512/
2. Alexander Tkachenko, https://publons.com/researcher/3041889/alexander-tkachenko/, https://stepik.org/course/94/


Most of the material updates will be published here.

Disclaimers:

1. All the course materials will be in English
2. By "gene expression" we almost always mean "RNA expression" if not said otherwise

## Syllabus

+ Central dogma of molecular biology, structure of gene, types of RNA, structure of RNA, transcription, reverse transcription, qPCR, FISH, <a href="https://konsolerr.github.io/gene_expression_2019/facs" target="_blank">FACS</a>
+ Microarray: quantification, normalization, basic analysis
+ RNA-seq: alignment, quantification, QC, normalization, basic analysis
+ Overall quality control: PCA, clustering, outlier detection
+ Overall quality control: batch correction
+ Differential expression (DE): limma for microarray, Deseq2 for RNA-seq
+ Downstream analysis: pathway/gene set enrichment analysis
+ Downstream analysis: gene expression deconvolution
+ Transcriptome assembly, functional annotation
+ Single-cell transcriptomics: Seurat basic analysis
+ Single-cell transcriptomics: Trajectory analysis, RNA velocity, optimal transport
+ Visual data exploration: phantasus, JBR genome browser
+ Experimental design of gene expression study

## Prerequisites

Most of the lessons will be in format of seminars: lectures and practical part with your laptops. Prior the start of the course, you are required to install latest version of R (https://cran.r-project.org/bin/) and Rstudio (https://www.rstudio.com/) to your laptop. Most of the practices we will be working in R and only for several we will require a command line.

## Homeworks 

There will be 4 projects homeworks during this course:

1. Microarray dataset: processing, QC, analysis, DE
2. RNA-seq dataset: processing (from fastq), QC, analysis, DE, downstream analysis (pathways)
3. Transcriptome assembly and functional annotation
4. Single-cell RNA-seq dataset: analysis with Seurat


## Below we will put links to slides and other useful links

### Lecture materials

1. Lecture 1: <a href="https://konsolerr.github.io/gene_expression_2019/intro" target="_blank">introduction</a> and <a href="https://konsolerr.github.io/gene_expression_2019/facs" target="_blank">FACS</a>
2. Lecture 2: <a href="https://konsolerr.github.io/gene_expression_2019/transcription/transcription.pdf" target="_blank">transcription</a>
3. Lecture 3: Microarrays <a href="https://konsolerr.github.io/gene_expression_2019/microarray/microarray.html" target="_blank">part 1</a> and <a href="https://konsolerr.github.io/gene_expression_2019/microarray/microarray2.html" target="_blank">part 2</a>
4. Lecture 4: Secondary analysis <a href="https://konsolerr.github.io/gene_expression_2019/secondary_analysis" target="_blank">slides</a> and <a href="https://konsolerr.github.io/gene_expression_2019/secondary_analysis/secondary_analysis.zip" target="_blank">materials</a>
5. Lecture 5: RNA-seq <a href="https://konsolerr.github.io/gene_expression_2019/rna-seq" target="_blank">slides</a> and <a href="https://konsolerr.github.io/gene_expression_2019/rna-seq/rna_seq_materials.zip" target="_blank">materials</a>
6. Lecture 6: Transcriptome assembly and annotation. <a href="https://konsolerr.github.io/gene_expression_2019/assembly/Transcriptome_assembly.pdf" target="_blank">Slides</a>
7. Lecture 7: gene expression deconvolution <a href="https://konsolerr.github.io/gene_expression_2019/deconvolution" target="_blank">slides</a> and <a href="https://konsolerr.github.io/gene_expression_2019/deconvolution/deconvolution.R" target="_blank">code</a>.
8. Lecture 8: RNA-seq normalization and DE <a href="https://konsolerr.github.io/gene_expression_2019/normalization" target="_blank">slides</a> and <a href="https://konsolerr.github.io/gene_expression_2019/normalization/normalization_materials.zip" target="_blank">materials</a>
9. Lecture 9: scRNA-seq <a href="https://konsolerr.github.io/gene_expression_2019/scRNA-seq" target="_blank">slides</a> and <a href="https://konsolerr.github.io/gene_expression_2019/scRNA-seq/materials.zip" target="_blank">materials</a>. Brief lecture about trajectory methods for scRNA-seq: <a href="https://konsolerr.github.io/gene_expression_2019/scRNA-seq/trajectory.pdf" target="_blank">slides</a>
10. Lecture 10: Visual data exploration. Phantasus, single-cell explorer, JBR genome browser: <a href="https://konsolerr.github.io/gene_expression_2019/visual/visual.pdf" target="_blank">slides</a> and <a href="https://konsolerr.github.io/gene_expression_2019/visual/human_cd14_reads_cov.yaml" target="_blank">human_cd14_reads_cov.yaml</a>

Random materials:

1. Good chapter on SVD: <a href="https://www.cs.cmu.edu/~venkatg/teaching/CStheory-infoage/book-chapter-4.pdf" target="_blank">https://www.cs.cmu.edu/~venkatg/teaching/CStheory-infoage/book-chapter-4.pdf</a>
2. Good interactive material "Understanding UMAP": <a href="https://pair-code.github.io/understanding-umap/" target="_blank">https://pair-code.github.io/understanding-umap/</a>
3. How much scRNA-seq library prep costs: https://satijalab.org/costpercell
4. About distributions behind box plots: https://www.autodeskresearch.com/node/3699


### Homework 1 

The first homework is to analyze a microarray gene expression dataset. All the datasets will be available by GEO id, so you can find them at GEO omnibus and also use `GEOquery::getGEO("GSEXXXXX")` to get the expression data into R session.

Basically you needed to perform basic steps for a dataset: normalize and make sense of the PCA plot. Then you need to disgn differential exression to address the general question for your homework. Once you perform DE I want you to highlight genes that are different and perform pathway enrichment.

Datasets IDs are listed in the table below:

https://docs.google.com/spreadsheets/d/15tEZ3fADDdxzIxdBfFhAozQUMb-chAjX14qtdRJKtP8/edit?usp=sharing

### Homework 2

In the second homework you will be given a small transcirptome to assemble. You can use either your laptop (if it is powerfull enough) or a computational cluster that was set up for you. You can also use any other online tools (like galaxy or gVolante). To pass this homework you must present your results and possibly compare to paper results.

Datasets are:

1. Liu N, Zhang G, Xu S, Mao W et al. Comparative Transcriptomic Analyses of Vegetable and Grain Pea (Pisum sativum L.) Seed Development. Front Plant Sci 2015;6:1039. PMID: 26635856 https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE72573 https://www.ncbi.nlm.nih.gov/sra?term=SRP063052
2. Rai A, Yamazaki M, Takahashi H, Nakamura M et al. RNA-seq Transcriptome Analysis of Panax japonicus, and Its Comparison with Other Panax Species to Identify Potential Genes Involved in the Saponins Biosynthesis. Front Plant Sci 2016;7:481. PMID: 27148308 https://www.ncbi.nlm.nih.gov/sra?term=SRP071142 https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE78893

Analysis

* Run Trimmomatic (http://www.usadellab.org/cms/?page=trimmomatic) to filter low-quality reads
* Run Trinity (https://github.com/trinityrnaseq/trinityrnaseq/wiki)/rnaSPAdes (http://cab.spbu.ru/software/rnaspades/) to assemble the transcriptome
* Run BUSCO (https://busco.ezlab.org/) and TransRate (https://hibberdlab.com/transrate/index.html) to assess the transcriptome
* Optional: run cd-hit-est (http://weizhongli-lab.org/cd-hit/) to collapse similar transcripts
* Annotate with blast+ (ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/) using Uniprot SwissProt (can be downloaded here https://www.uniprot.org/downloads) as a reference

More details about datasets and cluster access are at https://docs.google.com/spreadsheets/d/15tEZ3fADDdxzIxdBfFhAozQUMb-chAjX14qtdRJKtP8/edit#gid=1796960723

### Homework 3

For third homework you will need to chose one of recent scRNA-seq datasets and perform basic analysis:

1) Preprocessing: filtering out bad cells and normalization
2) UMAP + clustering
3) Marker selection  for clusters

I recommend using code snippets from the lecture or usual seurat tutorial at https://satijalab.org/seurat/v3.1/pbmc3k_tutorial.html

Dataset suggestions (you can analyze any dataset you want, these are just some suggestions):

1. Transcriptome Analysis Reveals Nonfoamy Rather Than Foamy Plaque Macrophages Are Proinflammatory in Atherosclerotic Murine Models, https://www.ahajournals.org/doi/pdf/10.1161/CIRCRESAHA.118.312804. Link to the dataset: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM3215435
2. Suppressing miR-21 activity in tumor-associated macrophages promotes an antitumor immune response, https://www.jci.org/articles/view/127125. Link to the dataset: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE118931
3. Coopted temporal patterning governs cellular hierarchy, heterogeneity and metabolism in Drosophila neuroblast tumors, https://elifesciences.org/articles/50375. Link to the dataset: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM3161971

You will pass this homework by presenting results and comparing with the publication.
