{
  "cells": [
    {
      "cell_type": "raw",
      "metadata": {},
      "source": [
        "---\n",
        "title: \"Recent developments in scNavigator\"\n",
        "author: \"Konstantin Zaitsev\"\n",
        "date: \"14/09/2022\"\n",
        "format: \n",
        "  revealjs: \n",
        "    slide-number: true\n",
        "    theme: [libs/itmo.scss]\n",
        "    footer: \"Artyomov Lab Meeting (Sep 14th 2022)\"\n",
        "editor: source\n",
        "---"
      ],
      "id": "0d2f1583"
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## scNavigator\n",
        "\n",
        "[https://artyomovlab.wustl.edu/scn/](https://artyomovlab.wustl.edu/scn/)\n",
        "\n",
        "![](images/schematic.png)\n",
        "\n",
        "## scNavigator\n",
        "\n",
        "-  Web tool for visual analysis of scRNA-seq data: gene expression plots, pathway expression plots, clustering and so on\n",
        "-  Database of uniformly processed public scRNA-seq datasets\n",
        "-  Gene signature search / single gene search through public scRNA-seq datasets\n",
        "\n",
        "## scNavigator (code base)\n",
        "\n",
        "- scNavigator - [https://github.com/ctlab/scNavigator](https://github.com/ctlab/scNavigator). \n",
        "\n",
        "Kotlin server side + a lot of JavaScript client side.\n",
        "\n",
        "\n",
        "- scn-pipeline - [https://github.com/ctlab/scn-pipeline](https://github.com/ctlab/scn-pipeline). \n",
        "\n",
        "Snakemake pipeline for processing public datasets.\n",
        "\n",
        "\n",
        "- scnPrep - [https://github.com/ctlab/SCNPrep](https://github.com/ctlab/SCNPrep). \n",
        "\n",
        "R package to convert Seurat objects to format of scNavigator.\n",
        "\n",
        "\n",
        "## scn-pipeline\n",
        "\n",
        "-  Spring/Summer 2022\n",
        "-  Huge rework of the old code base from Maria Firuleva\n",
        "-  Old: custom scripts that would generate analysis pipeline specific for the platform (Slurm or LSF)\n",
        "-  Currently: one snakemake pipeline that does everything (you can use slurm or LSF profiles to run the pipeline)\n",
        "\n",
        "## scn-pipeline\n",
        "\n",
        "-  Github Repo: [https://github.com/ctlab/scn-pipeline](https://github.com/ctlab/scn-pipeline)\n",
        "-  Documentation (builds automatically from `docs` folder): [https://scn-pipeline.readthedocs.io/en/latest/](https://scn-pipeline.readthedocs.io/en/latest/)\n",
        "\n",
        "Currently we only support 10X and DropSeq, but in further version I will start implementing other technologies.\n",
        "\n",
        "## scn-pipeline, how to run\n",
        "\n",
        "Put the dataset you want to process in the `config/dataset.json` file as a list of datasets, example below.\n",
        "\n",
        "\n",
        "```{json}\n",
        "#| eval: false\n",
        "[\"GSE145241\", \"GSE116240\"]\n",
        "```\n",
        "\n",
        "\n",
        "and then simply execute two rules (example for local execution):\n",
        "\n",
        "\n",
        "```{bash}\n",
        "#| eval: false\n",
        "snakemake -j 4 --use-conda get_all_meta\n",
        "snakemake -j 4 --use-conda process_all\n",
        "```"
      ],
      "id": "b1c46e29"
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3",
      "language": "python",
      "name": "python3"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}